### PR TITLE
Background store

### PR DESCRIPTION
**Pull request by fyang1024**

### CODE REVIEW

File: src/ts/services/keyring-vault-proxy.ts [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+  })
+}
+export function unlockWallet (password: string, addressPrefix?: string):
+  Promise<Array<KeyringPair$Json>> {
+  const port = browser.runtime.connect(undefined, { name: '__SPECKLE__' })
+  return new Promise<Array<KeyringPair$Json>>(resolve => {
+    port.onMessage.addListener((msg) => {
+      if (msg.method === 'unlock') {
+        resolve(msg.result)
+        port.disconnect()
```
Comment (ttaoS) [inline src/ts/services/keyring-vault-proxy.ts]: should this be called before resolve?

File: src/ts/services/keyring-vault-proxy.ts [hunk 2]
```diff
@@ -0,0 +1,174 @@
+import { browser } from 'webextension-polyfill-ts'
+import { KeyringPair$Json } from '@polkadot/keyring/types'
+
+export function isWalletLocked (): Promise<boolean> {
+  const port = browser.runtime.connect(undefined, { name: '__SPECKLE__' })
```
Comment (mytechtip) [inline src/ts/services/keyring-vault-proxy.ts]: Is it possible to connect once and then the connection can be shared without disconnecting?

File: src/ts/background/index.ts [hunk 3]
```diff
@@ -13,3 +15,103 @@ wrapStore(store, {
...
+      case 'importAccountFromMnemonic':
+        try {
+          keyringVault.importAccountFromMnemonic(msg.mnemonic, msg.accountName).then((pairJson) => {
+            port.postMessage({ method: 'importAccountFromMnemonic', result: pairJson })
+          })
+        } catch (e) {
+          port.postMessage({ method: 'importAccountFromMnemonic', error: e })
+        }
+        break
+      case 'importAccountFromJson':
```
Comment (ttaoS) [inline src/ts/background/index.ts]: minor: func names can be put in const shared with background

**Comment by fyang1024** [comment]
Refactored.

File: src/ts/background/index.ts [hunk 4]
```diff
@@ -13,3 +16,106 @@ wrapStore(store, {
...
+  port.onMessage.addListener(function (msg) {
+    switch (msg.method) {
+      case FUNCS.IS_LOCKED:
+        port.postMessage({ method: FUNCS.IS_LOCKED, result: keyringVault.isLocked() })
+        break
+      case FUNCS.IS_UNLOCKED:
+        port.postMessage({ method: FUNCS.IS_UNLOCKED, result: keyringVault.isUnlocked() })
+        break
+      case FUNCS.UNLOCK:
+        keyringVault.lock()
```
Comment (ttaoS) [inline src/ts/background/index.ts]: does it need `await` before posting message?

File: src/ts/background/index.ts [hunk 5]
```diff
@@ -13,3 +16,106 @@ wrapStore(store, {
...
+          keyringVault.updateAccountName(msg.address, msg.accountName).then((pairJson) => {
+            port.postMessage({ method: FUNCS.UPDATE_ACCOUNT_NAME, result: pairJson })
+          })
+        } catch (e) {
+          port.postMessage({ method: FUNCS.UPDATE_ACCOUNT_NAME, error: e })
+        }
+        break
+      case FUNCS.REMOVE_ACCOUNT:
+        try {
+          keyringVault.removeAccount(msg.address)
```
Comment (ttaoS) [inline src/ts/background/index.ts]: no message required?

File: src/ts/background/index.ts (hunk 4 above)
Comment (fyang1024) [reply src/ts/background/index.ts]: I don't think so. We just need to make sure the listener is added before sending the message to background, right?

File: src/ts/background/index.ts (hunk 5 above)
Comment (fyang1024) [reply src/ts/background/index.ts]: No, just like unlock, if it's unlocked, doing nothing. if nothing removed, doing nothing. always success, always resolve

File: src/ts/background/index.ts (hunk 4 above)
Comment (ttaoS) [reply src/ts/background/index.ts]: i meant when popup page send unlock message. do we want to wait until `keyringVault.lock()` is completed then post message back. we may also want to try/catch here

Comment (ttaoS) [reply src/ts/background/index.ts]: don't worry about it. lock is not an async

Comment (fyang1024) [reply src/ts/background/index.ts]: Fixed typo, it should FUNCS.LOCK here

File: src/ts/background/index.ts (hunk 5 above)
Comment (ttaoS) [reply src/ts/background/index.ts]: If there is no message sent after removing account, how does proxy know it should be resolved? 

`
  port.onMessage.addListener((msg) => {
      if (msg.method === FUNCS.REMOVE_ACCOUNT) {
        resolve()
      }
    })
    port.postMessage({ method: FUNCS.REMOVE_ACCOUNT, address: address })
  })`

Comment (fyang1024) [reply src/ts/background/index.ts]: yes, you are right. Added missing message just now

**Pull request merged**